### PR TITLE
[pvr] check timer type exists before access

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

If a backend changes the timer types it supports this can mean we load timers without a valid type which would cause a seg fault.

## Motivation and Context

Seg faults!

## How Has This Been Tested?

OSX with pvr.vuplus.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

@ksooo if you have time to review. Is there a better approach?

In PVRDatabase.cpp only the first two lines are relevant. The rest are clang format.

### CODE REVIEW

File: xbmc/pvr/PVRDatabase.cpp [hunk 1]
```diff
@@ -977,34 +977,49 @@ bool CPVRDatabase::Persist(CPVRTimerInfoTag& timer)
 
   // insert a new entry if this is a new timer, or replace the existing one otherwise
   std::string strQuery;
+  const std::shared_ptr<CPVRTimerType> timerType = timer.GetTimerType();
+  unsigned int timerTypeId = timerType ? timerType->GetTypeId() : PVR_TIMER_TYPE_NONE;
```
Comment (ksooo) [inline xbmc/pvr/PVRDatabase.cpp:981]: Only local timers (== reminders) should be persisted in the database. And the timer type of a reminder never must be nullptr. I fear we have a bug somewhere that leads to storing non-reminder timers in the db.

**Comment by ksooo** [comment]
> If a backend changes the timer types it supports
 
This must not happen. Our code is not prepared to handle this and (as you can see with your segvault) this can lead to any kind of errors. For example, only local timers (== reminders) should ever be stored in the db. The fact you needed to fix the db code is a consequence of our code not being ready for timers without a type.

The right fix imo is, that In case a backend timer type changes, your addon must trigger a timers update so that Kodi reloads the timers.

**Comment by ksooo** [comment]
Wait, it's even worse. We're loading the timer types from the client only once after 'connect' and cache it, then (see  CPVRClient::GetTimerType implementation). So, just triggering a timer reload is not sufficient and there is no API to trigger a timer types reload... something we might need to add for v20.

**Comment by phunkyfish** [comment]
Ok, so what’s the correct fix for v19?

**Comment by ksooo** [comment]
Maybe we should refetch timer types on every add-on l-triggered timer update? First refetch types, then timers.

**Comment by phunkyfish** [comment]
Sounds reasonable. Although this PR would still be required. We should protect against a bad timer type id being sent from an add-on regardless.

**Comment by ksooo** [comment]
I'm not a fan of hiding bugs by adding null pointer protection for pointers which must not be null, which this PR in fact does.

Storing timers without a timer type in the DB is wrong. Just making it not crash is no solution.

**Comment by phunkyfish** [comment]
The issue is not that timers without a type are stored in the DB. At the time they were stored they had a type. Right, or an
I following this incorrectly.

**Comment by phunkyfish** [comment]
I guess these should be removed from the DB as they are not valid.

**Comment by ksooo** [comment]
> At the time they were stored they had a type.

No timers created by backends are ever supposed to be stored in the db, only reminders which are always local timers created by Kodi. The problem you have faced is that a backend timer which formerly had a type shall be now stored in the db (for the first time), because the PVR core logic to identify whether a timer is owned by a backend fails if no timer type is present. See implementation of CPVRTimerInfoTag::IsOwnedByClient.

**Comment by ksooo** [comment]
CPVRTimerInfoTag::IsOwnedByClient() is the actual problem. Gimme a sec to think about the right fix, please...

**Comment by ksooo** [comment]
> I guess these should be removed from the DB as they are not valid.

 No, they should **never** get into the db!

**Comment by ksooo** [comment]
Okay, one step back: Where come those CPVRTimerInfoTag instances not having an m_timerType come from?

As a fact, those instances are not loaded from the PVR db. Please confirm. If they were stored in the db, we have to find another bug, because backend created timers are not supposed to be stored in the PVR db.

If the instances do not come from the db, the addon must have returned PVR_TIMER structs with an invalid timer type, means a type, that is not included in addon's GetTimerTypes result. This is a bug. Those PVR_TIMERs never should be returned to Kodi.

**Comment by ksooo** [comment]
In case you are still able to reproduce the original issue:

1) remove this PR from yur xbmc sources
2) delete TV37.db
3) Set a breakpoint here: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/timers/PVRTimerInfoTag.cpp#L74
4) Set a breakpoint here: https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/timers/PVRTimerInfoTag.cpp#L157
=> if you hit one of the breakpoints, look back where this faulty timer comes from and fix that place.

**Comment by phunkyfish** [comment]
I connected to a remote backend to debug a user crash. But then the timers from my backend caused the problem. If I had cleared the DB in between there would have been no issue.

**Comment by phunkyfish** [comment]
It was point 4 was the point I hit.

**Comment by phunkyfish** [comment]
Correction, I'm not sure it was either. Here is the log statement I hit: `CreateFromIds: Unable to resolve numeric timer type (0, -1)`

**Comment by ksooo** [comment]
Sorry, I don't get it. If we ever encounter a CPVRTimerInfoTag instance without an m_timerType instance, this is definitely a bug, because the timer types given by the addon do not match the types used in the PVR_TIMER structs returned by the same addon. 

Adding m_timertype nullptr checks does not really fix anything. In fact I should remove all existing m_timertype nullptr checks, and add an asstert() in the CPVRTimerInfotag ctors. I'm serious about that.

**Comment by phunkyfish** [comment]
Somehow these timers are getting into the DB. Because that is where they are loaded from. I wish I had kept a backup of the DB now.

**Comment by ksooo** [comment]
We need to reproduce first, how the timers are getting into the db. Then we can discuss about the fix. :-)

**Comment by phunkyfish** [comment]
I'm trying to figure out how they in there in the first place. Facepalm, they only went in there cause of my PR :(

**Comment by ksooo** [comment]
> Facepalm, they only went in there cause of my PR :(

 Yes, this is what I wanted to tell you the whole time... ;-)

**Comment by ksooo** [comment]
I still think the problem is your (?) addon created bogus PVR_TIMER instances. You must not put in invalid timer type ids, by means not contained in GetTimerTypes return values.

**Comment by phunkyfish** [comment]
Thanks for digging me out the hole I dug myself. I had actually already fixed it in the add-on but was scratching my head cause it was still crashing: https://github.com/kodi-pvr/pvr.vuplus/pull/309

**Comment by phunkyfish** [comment]
Closing as this PR is pointless.

**Pull request closed without merging**